### DISCUSSION TITLE
Best way to create a native DirectX swapchain Avalonia control?

### DISCUSSION THREAD

**Discussion by Sergio0694** [AvaloniaUI/Avalonia#12105] [Q&A]
Hey everyone 👋
I'm thinking about potentially creating a `ComputeSharp.Avalonia` library (an extension for https://github.com/Sergio0694/ComputeSharp), which would basically have the same controls I have for the UWP and WinUI 3 libraries (ie. `ComputeShaderPanel` and `AnimatedComputeShaderPanel`), but for Avalonia. I've already built a demo prototype (https://github.com/Sergio0694/ComputeSharp/tree/dev/avalonia) which does seem to work just fine:

![](https://github.com/AvaloniaUI/Avalonia/assets/10199417/4d0a6f97-0129-4dd1-a7d0-b548a306b61a)

For this test, I'm just using a `NativeControlHost` and then creating an `HWND` swapchain on it. I was wondering — is this the best way to go, or is there some built-in support for swapchain-based custom controls I should rather be using (kinda like `SwapChainPanel` on UWP/WinUI 3?). Or, now with all the new composition improvements in Avalonia 11, is there a better way to integrate something like this in a custom control, maybe even leveraging the built-in Windows compositor and using a swapchain for composition instead (not `HWND`)? 🤔

cc. @MikeCodesDotNET maybe you know, or would know who I should ask? Thank you! 😄

**Comment by hez2010** [comment]
> is there some built-in support for swapchain-based custom controls I should rather be using

Yes!
I recently integrated my renderer which is implemented in DirectX 11 into Avalonia, using the build-in GPU Interop feature in Avalonia.
Here is the repo for reference: https://github.com/hez2010/VolumeRenderer
Also, there is a sample from Avalonia: https://github.com/AvaloniaUI/Avalonia/tree/master/samples/GpuInterop/D3DDemo
Hope it helps.

**Comment by morphinapg** [comment]
I would like Avalonia support for ComputeSharp. Are you still considering doing this?
